### PR TITLE
Validate format of DUNs number

### PR DESCRIPTION
**Pull request by jessieay**
* Validate that it contains only integers
* Validate that it is 8,9, or 13 chars long (` Samwise::Util.format_duns(duns: value)`)

Closes https://github.com/18F/micropurchase/issues/425

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -1,4 +1,5 @@
 class User < ActiveRecord::Base
   has_many :bids
   validates :email, email: true, allow_blank: true
+  validates :duns_number, format: { with: /\A\d{8,13}\z/ }
```
Comment (adelevie) [inline app/models/user.rb]: https://github.com/18F/samwise/tree/micropurchase-compatibility#validate-the-format-of-a-duns-number

^ could be used instead

**Comment by mtorres253** [comment]
The second validation was not in the story. And not entirely correct. The number should be 8, 9 or 13 digits (so not 10, 11 or 12 digits, for example). There is another story for this that has to be prioritized. However, I'm wondering how much more validation we will need if we can successfully tell the user whether their DUNS number exists in Sam.gov or not.

**Comment by jessieay** [comment]
@mtorres253 

In this validation we are calling `Samwise::Util.format_duns(duns: '08-011-5718')`, which removes any dashes and adds zero padding if the DUNs number is 8 or 9 chars: https://github.com/18F/samwise/tree/micropurchase-compatibility#format-a-duns-number

If the number submitted is 10, 11, or 12 digits long, the samwise lib raises a `Samwise::Error::InvalidFormat` error and we return our error msg.
 
If the DUNs is the correct # of chars (8, 9, or 13), we then make sure that they are all integers. Unfortunately, the samwise lib does not do this for us so we are doing it here with a regex.

re: 

> wondering how much more validation we will need if we can successfully tell the user whether their DUNS number exists in Sam.gov or not.

I think this validation is helpful in the case that someone submits a typo or something.

File: spec/models/update_user_spec.rb [hunk 2]
```diff
@@ -45,10 +46,44 @@
...
+          }
+        )
+
+        updater = UpdateUser.new(params, current_user)
+        updater.save
+
+        expect(updater.errors).to eq(
+          'DUNS number format is invalid'
+        )
+      end
```
Comment (harrisj) [inline spec/models/update_user_spec.rb]: Can we also check that sam_account is set to false (even if it was true?)

**Comment by harrisj** [comment]
This mostly looks good to me, but I think you're going to need to rebase it

File: spec/models/update_user_spec.rb (hunk 2 above)
Comment (jessieay) [inline spec/models/update_user_spec.rb]: ok!

Comment (jessieay) [inline spec/models/update_user_spec.rb]: ah wait -- this is not hitting the SAM API -- we are validating the format. So we are not actually saving this new duns when someone submits it (we keep old value until they submit a potentially valid duns number, at which point we hit the API.

added spec expectation to clarify this behavior at 20b605d166498e1aace44de2668bd342f72e5103

**Comment by jessieay** [comment]
@harrisj rebased!

File: spec/models/update_user_spec.rb [hunk 3]
```diff
@@ -45,10 +46,46 @@
...
+          user: {
+            duns_number: nil
+          }
+        )
+
+        updater = UpdateUser.new(params, current_user)
+        updater.save
+
+        expect(updater.errors).to eq('')
+      end
```
Comment (harrisj) [inline spec/models/update_user_spec.rb]: Should this also include a test that sam_account is reset to false?

File: app/models/authenticator.rb [hunk 4]
```diff
@@ -16,7 +16,9 @@ def find_or_create_user
   end
 
   def update_user
-    user.update_attribute(:name, name) if user.name.blank?
+    if user.name.blank?
+      user.update(name: name)
+    end
```
Comment (harrisj) [inline app/models/authenticator.rb]: I'm fine with this, but we should probably change the Rubocop rules in another PR to be more lenient as well

Comment (jessieay) [inline app/models/authenticator.rb]: I refer to avoid trailing conditionals; is rubocop mad at me?

File: spec/models/update_user_spec.rb (hunk 3 above)
Comment (jessieay) [inline spec/models/update_user_spec.rb]: we have a test for that in `spec/models/sam_account_reckoner_spec.rb` (L65) -- seems that is more the concern of the reckoner.

**Comment by jessieay** [comment]
Closing in favor of https://github.com/18F/micropurchase/pull/445

**Pull request closed without merging**